### PR TITLE
Keep NMS index gathering on cuda device

### LINKED ISSUE #8713: `torchvision.ops.boxes.batched_nms` slow on large box numbers
## Description

`torchvision.ops.boxes.batched_nms` on CUDA GPU slows down considerably when then number of bounding boxes involved increases.

The slow down is associated with Device -> Host transfer, and is linked to the iterative part of the Non Maximum Suppression (NMS) algorithm. In a nutshell the IoU map is computed on the device, then the mask is copied to the CPU to perform the iterative unwrap, which result is copied back to the device (from [here and below](https://github.com/pytorch/vision/blob/868a3b42f4bffe29e4414ad7e4c7d9d0b4690ecb/torchvision/csrc/ops/cuda/nms_kernel.cu#L136)).

The mask size grows quadratically with the number of input bounding boxes and we see a large TX rate when running on 30_000+ boxes.

In comparison the [OpenLabs mmcv](https://github.com/open-mmlab/mmcv) solution does the same thing for the IoU map but runs a custom kernel to do the unwrap directly on the device. The[ implemented kernel](https://github.com/open-mmlab/mmcv/blob/71437a361cc8918fc398ae408267cf019f4ca03f/mmcv/ops/csrc/common/cuda/nms_cuda_kernel.cuh#L76) is not very efficient compute wise but save the data transfer cost, which is the main bottleneck.

I benchmarked `torchvision` batched_nms against `mmcv`'s on `V100` and `A100` GPUs.
A100_bench_rel_loglog
V100_bench_rel_loglog
Both figures show the speed factor when comparing a solution to `torchvision.ops.boxes._batched_nms_vanilla` (there is 2 nms in torchvision, selected based on the number of elements. Here , `torchvision.ops.boxes._batched_nms_vanilla` is used a base comparison and we compare `torchvision.ops.boxes._batched_nms_coordinate_trick` and `mmcv` batched_nms). From 30k boxes and above `mmcv` NMS is x20+ faster.

Is there a reason why we keep this GPU -> CPU transfer ?
Could we improve the scalability by having a similar on-device additional kernel ?

## Additional informations

* All boxes are from the same class
* Benchmark has been done using `torch.utils.benchmark.Timer` on 100 examples for each NMS.
* I did not know if this should be put as Bug report or Feature request.

### Versions

```
PyTorch version: 2.5.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.24.1
Libc version: glibc-2.35

Python version: 3.10.14 (main, May 14 2024, 06:11:20) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.10.219-208.866.amzn2.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 535.183.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.2
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               96
On-line CPU(s) list:                  0-95
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8275CL CPU @ 3.00GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   2
Core(s) per socket:                   24
Socket(s):                            2
Stepping:                             7
BogoMIPS:                             5999.99
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            1.5 MiB (48 instances)
L1i cache:                            1.5 MiB (48 instances)
L2 cache:                             48 MiB (48 instances)
L3 cache:                             71.5 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-23,48-71
NUMA node1 CPU(s):                    24-47,72-95
Vulnerability Gather data sampling:   Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:          KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Vulnerable
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.1.2
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] onnx==1.17.0
[pip3] onnxruntime==1.20.0
[pip3] pytorch-ranger==0.1.1
[pip3] torch==2.5.0
[pip3] torch-optimizer==0.3.0
[pip3] torchaudio==2.5.0
[pip3] torchmetrics==1.4.0.post0
[pip3] torchvision==0.20.0
[pip3] triton==3.1.0
[conda] No relevant packages
```

### PR DESCRIPTION
**Pull request by Ghelfi**
Performs the unwrap of IoU mask directly on the cuda device in NMS.

This prevents device -> host -> device data transfer. 

fixes #8713

### CODE REVIEW

**Comment by CNOCycle** [comment]
I would like to share an important finding regarding the root cause of slow performance in NMS when dealing with a large number of boxes (#8713). The issue is primarily due to the occurrence of minor page faults, as detailed in [1]. When data is transferred from the GPU to the host, the operating system must locate an appropriate memory space to store these temporary variables, and this operation can be quite costly in terms of execution time. As shown in Fig. 2 of [1], execution time curves diverge into two distinct groups as the number of objects increases, with varying results across different hardware configurations. Further analysis is provided in Table 1 and Section **Performance Analysis** of [1].

To summarize, the key takeaways are:
* The slow execution time of NMS can be reproduced on both x86-64 and ARM systems, across various generations of Nvidia GPU combanations.
* To reduce the execution time of NMS, we should aim to minimize the occurrence of minor page faults.

Since end users typically do not have the privilege to modify the operating system, and adjusting the lifecycle of these temporary variables may fall outside the scope of `torchvision`, I suggest the following approaches, which are detailed in [1]:

1. **CPU-free NMS**: This method is the same as the approach proposed in PR (#8766), and its mechanism has been extensively studied.
2. **Async NMS**: The performance of Async NMS depends on various factors, including the version of CUDA, GPU driver, the operating system, and current memory usage. More details can be found in the **Overhead of Minor Page Faults** section of the **Discussion** in [1]. 
3. **Hybrid NMS**: This approach is more complex, requiring meta-information about the currently used system. It is thus more suitable for advanced users.

The experimental comparison among three approaches can be found in [1]. Personally, **I highly recommend CPU-free NMS.** Nevertheless, the simplest implementation of Async NMS could potentially provide performance benefits if data copy is set to non-blocking mode, along with adding system synchronization before the CPU accesses the data:
```c++
//  nms_kernel_impl on GPU
at::Tensor mask_cpu = mask.to(at::kCPU, /* non_blocking*/ true);
// some non-relevant CPU codes
cudaStreamSynchronize(stream)
// unsigned long long* mask_host = (unsigned long long*)mask_cpu.data_ptr<int64_t>();
```

The following is result of default NMS and CPU-free NMS executed on V100 with the latest docker image `nvcr.io/nvidia/pytorch:24.12-py3`. It can be seen that the execution time of CPU-free NMS is half of the default NMS.
fig_nms_objs_scatter_default
fig_nms_objs_scatter_gpu




**I would strongly recommend that the community can cite the paper [1]** to help users understand that the code modifications involved in the PR (#8766) are based not just on experimental results, but also on clear explanations and insights drawn from computing architecture.

References:

[1] [Chen, E. C., Chen, P. Y., Chung, I., & Lee, C. R. (2024). Latency Attack Resilience in Object Detectors: Insights from Computing Architecture. In Proceedings of the Asian Conference on Computer Vision (pp. 3206-3222).](https://openaccess.thecvf.com/content/ACCV2024/papers/Chen_Latency_Attack_Resilience_in_Object_Detectors_Insights_from_Computing_Architecture_ACCV_2024_paper.pdf)

**Comment by antoinebrl** [comment]
Well done, the improvement looks quite significant ! @NicolasHug any chance this make it to the next release of `torchvision` ?

**Comment by CNOCycle** [comment]
To provide a proper NMS implementation for end users, I would like to offer the following personal comments.

As I mentioned earlier, my NMS-free NMS implementation differs slightly from the proposed one. In my approach, the function `gather_keep_from_mask` is fused into `nms_kernel_impl`, which results in only one GPU kernel being launched. Also, **this kernel is executed by a single warp** to ensure correctness.

From what I understand, most object detectors limit the top-k objects fed into the NMS function, with k generally set between 1000 and 5000. In this range, the elapsed time for NMS is under 2 ms in both the random and best-case scenarios. (The worst case is unlikely to occur with real-world datasets.) However, the overhead of launching a second GPU kernel is approximately 1-5 ms, which depends heavily on the CPU's status.

That being said, I completely agree that executing `nms_kernel_impl` with only one warp becomes inefficient when dealing with more than 20,000 objects. Even in a suboptimal CPU-free NMS implementation, its performance exceeds that of Async NMS. For this reason, I highly recommend replacing the default NMS implementation with the CPU-free NMS.

In summary, I believe that feeding 20,000+ objects into NMS is a rare occurrence, and the fused NMS implementation can offer slightly better performance in most scenarios.

Another issue to consider is whether we should maintain two NMS implementations in `torchvision` (`_batched_nms_coordinate_trick` and `_batched_nms_vanilla`) once the performance of the NMS kernel has been refined.

**Comment by Ghelfi** [comment]
@NicolasHug Can we try to ship this with the next release if we decide to move forward.

**Comment by NicolasHug** [comment]
Yes, I'll try to make this happen for the next release. Sorry, time has been scarce lately.

**Review by NicolasHug — APPROVED** [review]
Thanks a lot for the PR.

In all honesty I didn't meticulously checked the correctness of the new kernel - I'm mostly relying on our existing tests for that.

For my own record, I ran the following benchmarks where I could reproduce the massive improvements as `num_boxes` grows.

<details>

```py
import torch
from time import perf_counter_ns
from torchvision.ops import nms


def bench(f, *args, num_exp=100, warmup=0, **kwargs):

    for _ in range(warmup):
        f(*args, **kwargs)

    times = []
    for _ in range(num_exp):
        start = perf_counter_ns()
        f(*args, **kwargs)
        torch.cuda.synchronize()
        end = perf_counter_ns()
        times.append(end - start)
    return torch.tensor(times).float()

def report_stats(times, unit="ms", prefix=""):
    mul = {
        "ns": 1,
        "µs": 1e-3,
        "ms": 1e-6,
        "s": 1e-9,
    }[unit]
    times = times * mul
    std = times.std().item()
    med = times.median().item()
    print(f"{prefix}{med = :.2f}{unit} +- {std:.2f}")
    return med


def make_boxes(num_boxes, num_classes=4):
    boxes = torch.cat((torch.rand(num_boxes, 2), torch.rand(num_boxes, 2) + 10), dim=1).to("cuda")
    assert max(boxes[:, 0]) < min(boxes[:, 2])  # x1 < x2
    assert max(boxes[:, 1]) < min(boxes[:, 3])  # y1 < y2

    scores = torch.rand(num_boxes).to("cuda")
    idxs = torch.randint(0, num_classes, size=(num_boxes,)).to("cuda")
    return boxes, scores, idxs

NUM_EXP = 30
for num_boxes in (10, 100, 1000, 10000, 100000):
    boxes, scores, _ = make_boxes(num_boxes)
    times = bench(nms, boxes, scores, iou_threshold=.7, warmup=1, num_exp=NUM_EXP)
    report_stats(times, prefix=f"{num_boxes = } ")
```

```
On main
num_boxes = 10 med = 0.24ms +- 0.10
num_boxes = 100 med = 0.25ms +- 0.02
num_boxes = 1000 med = 0.31ms +- 0.02
num_boxes = 10000 med = 3.18ms +- 3.78
num_boxes = 100000 med = 1408.21ms +- 27.27

This PR
num_boxes = 10 med = 0.25ms +- 0.10
num_boxes = 100 med = 0.26ms +- 0.02
num_boxes = 1000 med = 0.29ms +- 0.03
num_boxes = 10000 med = 0.95ms +- 0.04
num_boxes = 100000 med = 15.32ms +- 1.50
```

</details>


I'll merge this PR when it's ready and probably will follow-up with some update to the batched-nms heuristic, as @CNOCycle  suggested.
